Implement high-performance identifiable key scanning in managed code

nguerrera opened the pull request:
This change removes the interop with the rust API from `IdentifiableScan` and implements similar high-performance scanning techniques in managed code. This reduces the size of the NuGet package tremendously, from ~7 MB to ~200 KB, and adds support for high-performance scanning on non x86/x64 CPUs and non-Windows operating systems.

Performance is generally better than before for managed callers of `IdentifiableScan` with significant gains in most cases. See benchmarks below. 

Code generation is used to avoid having to keep the high-performance scanning code in sync when patterns are changed. See also extensive code comments on the approach.

For now, the streaming capability of `IdentifiableScan` has been removed, but this can be re-added as the managed algorithm is actually streaming friendly by only ever regex-matching against short sub-ranges of input. This was held back for lack of time and discovery that our only partner that calls this API does not use the streaming. #155

Finally, the `StyleCop` analyzer is removed because it broke on newer C# constructs. This will be replaced with a `dotnet format` check like the one in `microsoft/cask` in a follow-up change.

# Benchmark

Scan all generated true positive examples of all identifiable patterns, with 100 KB of other data before each match.

`RandomPrefix` is random base64 data and `NonRandomPrefix` is all `&` characters. The former stresses the strength of signature sniffing algorithm which varies between .NET 8, .NET 9, .NET Framework, and Rust. The latter contains nothing that could be part of a signature except in a match, so just about any signature sniffing algorithm should perform well. We also loosely refer to these as the 'best' and 'worst' cases in notes below.

# Environment
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.26100.3194)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
```

# Summary of results

* .NET 9: ~21x faster on worst case, ~6.9x faster on best case.
* .NET 8: ~1.5x faster on worst case, ~8.4x faster on best case. 
* .NET Framework: ~2.4x faster on worst case, ~17x faster on best case.

😍

# Full results

## .NET 9

### Before this PR

| Method                                    | Mean        | Error    | StdDev   |
|------------------------------------------ |------------:|---------:|---------:|
| IdentifiableScan_RandomPrefix             |   189.72 ms | 1.931 ms | 1.806 ms |
| IdentifiableScan_NonRandomPrefix          |    61.35 ms | 0.999 ms | 0.935 ms |
| SecretMasker_CachedDotNet_RandomPrefix    |   264.72 ms | 5.072 ms | 4.981 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix |   263.14 ms | 1.877 ms | 1.755 ms |
| SecretMasker_RE2_RandomPrefix             | 1,031.28 ms | 9.023 ms | 7.998 ms |
| SecretMasker_RE2_NonRandomPrefix          | 1,026.15 ms | 3.668 ms | 3.251 ms |


### After this PR

| Method                                    | Mean         | Error     | StdDev    |
|------------------------------------------ |-------------:|----------:|----------:|
| IdentifiableScan_RandomPrefix             |     9.186 ms | 0.1749 ms | 0.1718 ms |
| IdentifiableScan_NonRandomPrefix          |     8.843 ms | 0.0890 ms | 0.0832 ms |
| SecretMasker_CachedDotNet_RandomPrefix    |   260.554 ms | 2.6871 ms | 2.3821 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix |   261.570 ms | 2.7378 ms | 2.5609 ms |
| SecretMasker_RE2_RandomPrefix             | 1,020.852 ms | 2.4914 ms | 2.2086 ms |
| SecretMasker_RE2_NonRandomPrefix          | 1,033.865 ms | 3.5205 ms | 3.2931 ms |


## .NET 8

### Before this PR

| Method                                    | Mean        | Error    | StdDev   |
|------------------------------------------ |------------:|---------:|---------:|
| IdentifiableScan_RandomPrefix             |   193.15 ms | 3.466 ms | 3.073 ms |
| IdentifiableScan_NonRandomPrefix          |    61.72 ms | 0.835 ms | 0.781 ms |
| SecretMasker_CachedDotNet_RandomPrefix    |   271.70 ms | 0.940 ms | 0.734 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix |   264.04 ms | 4.581 ms | 4.285 ms |
| SecretMasker_RE2_RandomPrefix             | 1,032.38 ms | 3.171 ms | 2.811 ms |
| SecretMasker_RE2_NonRandomPrefix          | 1,020.46 ms | 2.632 ms | 2.462 ms |

### After this PR

| Method                                    | Mean         | Error     | StdDev    |
|------------------------------------------ |-------------:|----------:|----------:|
| IdentifiableScan_RandomPrefix             |   128.917 ms | 1.0331 ms | 0.9158 ms |
| IdentifiableScan_NonRandomPrefix          |     7.374 ms | 0.1434 ms | 0.2103 ms |
| SecretMasker_CachedDotNet_RandomPrefix    |   273.672 ms | 3.2205 ms | 3.0125 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix |   259.952 ms | 2.7058 ms | 2.5310 ms |
| SecretMasker_RE2_RandomPrefix             | 1,025.887 ms | 4.9130 ms | 4.5956 ms |
| SecretMasker_RE2_NonRandomPrefix          | 1,018.437 ms | 3.9070 ms | 3.6546 ms |

## .NET Framework

### Before this PR

| Method                                    | Mean       | Error    | StdDev   |
|------------------------------------------ |-----------:|---------:|---------:|
| IdentifiableScan_RandomPrefix             |   706.1 ms |  4.52 ms |  4.23 ms |
| IdentifiableScan_NonRandomPrefix          |   679.0 ms |  3.65 ms |  3.42 ms |
| SecretMasker_CachedDotNet_RandomPrefix    | 4,262.5 ms | 29.07 ms | 25.77 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix | 5,038.9 ms | 10.80 ms | 10.10 ms |
| SecretMasker_RE2_RandomPrefix             | 5,052.6 ms | 13.99 ms | 12.40 ms |
| SecretMasker_RE2_NonRandomPrefix          | 5,877.4 ms | 51.33 ms | 48.02 ms |


## After this PR

| Method                                    | Mean        | Error     | StdDev    |
|------------------------------------------ |------------:|----------:|----------:|
| IdentifiableScan_RandomPrefix             |   291.25 ms |  3.845 ms |  3.596 ms |
| IdentifiableScan_NonRandomPrefix          |    39.15 ms |  0.226 ms |  0.200 ms |
| SecretMasker_CachedDotNet_RandomPrefix    | 4,221.73 ms |  6.341 ms |  5.932 ms |
| SecretMasker_CachedDotNet_NonRandomPrefix | 5,012.14 ms | 42.388 ms | 39.649 ms |
| SecretMasker_RE2_RandomPrefix             | 5,077.60 ms | 39.867 ms | 33.291 ms |
| SecretMasker_RE2_NonRandomPrefix          | 5,891.66 ms | 32.403 ms | 30.309 ms |


# Notes

- `SecretMasker` is unchanged in this PR while `IdentifiableScan` that used to use rust interop before this PR and is now implemented in managed code. One of the next steps will be to eliminate these competing API and have a single entry point that is as fast as possible for the patterns its given. #154.

- We should also apply similar signature sniffing techniques to more than identifiable keys: #156.

- "Before this PR" includes a bug-fix where IdentifiableScan wasn't using rust for HISv2 keys due to 'JQQJ9' vs 'JQQJ' signature mismatch bug. See #129 that was closed as it is subsumed by this PR. The actual numbers for IdentifiableScan without this fix are much worse, but not shown above.

- The fact that .NET 8 does slightly better than .NET 9 on the best case is actually expected. In this synthetic case, multi-char search is as good as multi-string search, and of course multi-char search is faster than multi-string search. The remarkable thing about the .NET 9 implementation is that it doesn't slow down as you throw it harder cases.

- These numbers are not representative of the rust library's performance in isolation. Much of the cost measured here is due to interop overhead, and much of the overhead is due to deficiencies in our own code and not intrinsic costs of interop in the CLR. Nevertheless, it would be a lot of work to tune the interop and then we'd still have the huge nupkg and non-portability. From other experiment not show here, even if the interop overhead were reduced to zero, we'd still not be materially faster in the best cases and we'd be even still be slower for some combinations of input and .NET version.

- It would be an interesting follow-up to make the benchmark have a parameter that is the probability of encountering a leading signature char capital like 'A', 'J', 'Q', '~', '+'. Currently we test RadomPrefix=P(3/64) and NonRandomPrefix=P(0). We could also try to determine what a realistic probability is for real data getting scanned in order to have a synthetic benchmark that is somewhat representative of the average case.

- What the heck is up with RE2 being slower, even on .NET Framework? #158 

- The pattern for making the high-performance regexes from the general regexes could be better: #151

rwoll commented on GeneratedRegexPatterns/HighConfidenceSecurityModels.json:
```diff
@@ -1,10 +1,10 @@
 [
   {
-    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?P<secret>[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{52}JQQJ9(?:9|D|H)[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890][A-L][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{16}[A-Za-z][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{7}(?:[abcde ...
+    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?<refine>[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{52}JQQJ9(?:9|D|H)[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890][A-L][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{16}[A-Za-z][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{7}(?:[abcdef ...
```
nit: is there a reason these regexes enumerate every character instead of e.g. `[a-zA-Z0-9]`? or even `[a-z\d]` (assuming case ins.)

Soemtimes regex are easier to read with explicit enumeration, but in this case, I find the verbose regexes mush harder to read as the key bits are burried in the noise.

nguerrera replied:
I don't know, and I wondered the same and I did replace this with ranges when making them into "high performance" regexes so that I could read those easier.

I consider this an unrelated, pre-existing issue, though, so I won't change it here.

michaelcfanning replied:
These patterns result from a well-intentioned failed experiment to look into building up more readable patterns by preferring readable interpolated string constants in code. The readability and maintainability of string literals that comprises regexes in the IDE has been a real problem in historical projects. As Nick has pointed out in the CASK project, C# helps with this issue by providing raw string literal support. We should file a ticket to clean up these expressions and leverage that feature if helpful.

The problem remain of how to effectively share elements of regex patterns effectively, there may not be a satisfying way to do that. Regexes are so flexible it's extremely easy for inconsistencies to materialize across rule kinds that are basically shared except for updated signatures etc. (for HIS stuff).

rwoll commented on GeneratedRegexPatterns/HighConfidenceSecurityModels.json:
```diff
@@ -1,10 +1,10 @@
 [
   {
-    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?P<secret>[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{52}JQQJ9(?:9|D|H)[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890][A-L][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{16}[A-Za-z][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{7}(?:[abcde ...
+    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?<refine>[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{52}JQQJ9(?:9|D|H)[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890][A-L][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{16}[A-Za-z][abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]{7}(?:[abcdef ...
```
There's a few other teams, IIRC, who consume these JSON files programmatically to do redaction in other langs (e.g. Python Azure CLI). Please coordinate with @michaelcfanning to confirm with the team that changing the capture group name (`secret` -> `refine`) won't impact them.

nguerrera replied:
@michaelcfanning Thoughts? Should I call this out in release notes?

nguerrera replied:
Also, is the change form `<?P` to `<?` OK?

michaelcfanning replied:
The `P` designation is actually important IIRC because RE2 requires this for retrieving groups. You've likely broken RE2 use of the pattern for this rule, so it's a nice catch.

I mentioned this elsewhere, as long as we don't have a clean solution for breaking apart the sensitive data, we should at least author invariant tests that insist we discover use of `(?P<refine)` named group in every pattern.

This will be a bit annoying for simple patterns that simply comprise a prefix/secret/suffix but it would be easily enforced in unit tests.

And yes, I'd document this change in the release notes.

nguerrera replied:
OK, all the other identifiable patterns didn't have the P so were already broken in this way. I thought I was adjusting this one to match the others, but it was the only one that was right apparently. 😅

I'll fix this and add the coverage.

michaelcfanning replied:
OK, so now I am suspicious of my own information. :) I *think* we have explicit use of RE2 in our tests and if these patterns were broken for named group retrieval, we should know. In other projects, there is some normalization of regexes to account for syntax differences in named groups, I'd need to check if it's in play here. Certainly we should be consistent! And if the absence of `P` verifiably works for RE2, that would be preferable. I can help double check this tomorrow, it's getting late for you!

nguerrera replied:
Yes, it appears RE2 supports both:

https://github.com/google/re2/blob/6dcd83d60f7944926bfd308cc13979fc53dd69ca/doc/syntax.txt#L53-L54

It appears that Python is the one that requires 'P'. P for python is how I will remember this.

https://docs.python.org/3/library/re.html

I further discovered that we are allocating a new copy of the pattern, removing 'P' on every call to GetOrCreateRegex. 😱 Time in and around GetOrCreateRegex shows up in profiling. #145 is part of the problem, this is also bad and standardizing on 'P' everywhere will make it worse.

If python is the main consumer of the .json, perhaps we should standardize on serialization only.

This is all getting quite far from this change :( I am going to do the thing that changes the least number of things in this PR and file an issue.

nguerrera replied:
Sigh, that wasn't easy, so I shaved the yak instead: #148

rwoll commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
 - FNS => False negative reduction in static analysis.
 
+# UNRELEASED
+- DEP: `System.Buffers` dependency has been removed. The package no longer has any dependencies.
```
🎉 zero-dependency packages are the best. Nice work on all of this!

rwoll commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
+# UNRELEASED
+- DEP: `System.Buffers` dependency has been removed. The package no longer has any dependencies.
+- BRK: .NET 6.0 and 7.0 are no longer supported as they have reached end-of-life. Use a [supported version version of .NET](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core): .NET 8.0 or greater. 
+- BRK:  .NET Framework 4.5.1 through 4.6.0 are no longer supported. Use a version of .NET Framework version that supports [.NET Standard 2.0](https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0): NET 4.6.1 or greater with .NET 4.7.2 or greater strongly recommended. Note that there are no supported versions of Windows that have a version of .NET Framework that woul ...
```
```suggestion
- BRK: .NET Framework 4.5.1 through 4.6.0 are no longer supported. Use a version of .NET Framework version that supports [.NET Standard 2.0](https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0): NET 4.6.1 or greater with .NET 4.7.2 or greater strongly recommended. Note that there are no supported versions of Windows that have a version of .NET Framework that would be affected by this change at runtime, but build changes may be required.
```

nit

rwoll commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
+- BRK:  .NET Framework 4.5.1 through 4.6.0 are no longer supported. Use a version of .NET Framework version that supports [.NET Standard 2.0](https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0): NET 4.6.1 or greater with .NET 4.7.2 or greater strongly recommended. Note that there are no supported versions of Windows that have a version of .NET Framework that woul ...
+- BRK: `IdentifiableScan` no longer supports stream input.
+- BUG: `SecretMasker` and `IdentifiableScan` considered non-alphanumeric delimiter preceding secret to be part of the match for `SEC101/200.CommonAnnotatedSecurityKey`.
+- BUG  `SEC101/061.LooseOAuth2BearerToken` had incorrect signatures, causing no matches to be found unless the input happened to also contain `sig=` or `ret=`
```
```suggestion
- BUG:  `SEC101/061.LooseOAuth2BearerToken` had incorrect signatures, causing no matches to be found unless the input happened to also contain `sig=` or `ret=`
```

nit

rwoll commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
+- BUG: `SecretMasker` and `IdentifiableScan` considered non-alphanumeric delimiter preceding secret to be part of the match for `SEC101/200.CommonAnnotatedSecurityKey`.
+- BUG  `SEC101/061.LooseOAuth2BearerToken` had incorrect signatures, causing no matches to be found unless the input happened to also contain `sig=` or `ret=`
+- PRF: `IdentifiableScan` did not use high-performance scanning techniques for `SEC101/178.AzureIotHubIdentifiableKey` and `SEC101/200.CommonAnnotatedSecurityKey`. This was not intended, but a bug triggered fall-back to slower scanning due to incorrect signatures being used to determine which patterns could be matched with high performance.
+- PRF: `IdentifiableScan` now implements high-performance scanning techniques in managed code. The performance has been found to be equal or better to the prior implementation via rust interop, with significant gains in some cases when using .NET 9.0. This also reduces the size of the NuGet package tremendously and adds support for non x86/x64 CPUs and non-Windows operating systems.
```
nit: since you do have quantifiable numbers already, consider replacing `tremendously` with the actual numbers. Celebrate your accomplishment!

rwoll commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -751,7 +751,11 @@ impl ScanOptions {
             }
 
-            if &data[52..58] != b"JQQJ99" {
+            if &data[52..57] != b"JQQJ9" {
```
Is there code coverage targeted at this line? Looks like this fixes an otherwise unnocticed bug.

nguerrera replied:
Hmm, this was covered when I first fixed it but now that we no longer interop it is not. I'm going to revert this rust change. We should do it separately after we settle everything about `9|D|H`.

rwoll commented on src/Microsoft.Security.Utilities.Core/PotentialSecurityKeys/SEC101_061_LooseOAuth2BearerToken.cs:
```diff
@@ -15,7 +15,6 @@ public OAuth2BearerToken()
             // https://datatracker.ietf.org/doc/html/rfc6750#section-2.1
             Pattern = @$"(?i)authorization:(\s|%20)bearer(\s|%20)(?<refine>[0-9a-z][{WellKnownRegexPatterns.UrlUnreserved}+\/=]*)([^{WellKnownRegexPatterns.UrlUnreserved}+/=]|$)";
-            Signatures = new HashSet<string>(new[] { "sig=", "ret=" });
         }
```
Thanks for fixing and i'm glad to see you re-enabled the commented out tests covering this.

rwoll commented on src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs:
```diff
@@ -30,15 +30,16 @@ public void SecretMasker_PreciselyClassifiedSecurityKeys_Detections()
     }
 
-    private void ValidateSecurityModelsDetections(IEnumerable<RegexPattern> patterns, bool preciseClassifications)
+    private void ValidateSecurityModelsDetections(IEnumerable<RegexPattern> patterns, bool preciseClassifications, bool allowAdditionalFindings = false)
```
Do we want to allow patterns to match multiple things?

On one hand, I think that's very useful and reasonable. E.g. I want to tag something as a JWT as well as more-specifically a GitHub JWT.

But does that imply the detection is not as precise as it can be?

I think the real fix here is to have rules have orders of precedence/priority. e.g. it's OK that Generic JWT is a superset of GitHub, but we just allow some fall through logic so we don't have to artifically try to awakwardly code the rules to be mutually exclusive themselves (which forces cyclic rule knowledge).

Not expecting to resolve my feedback/comment in this review as it's totally out of scope. Thinking aloud.

michaelcfanning replied:
I think for our current scenarios we will want to move something like a chain-of-responsibility pattern, where regexes are called in sequence and the first (most precise) match wins and suspends analysis. Or at least this library should be configurable for this approach. This is well suited to value classification, where we obtain discrete elements for scanning. In other scenarios, the data itself might simply be productive of multiple matches.

nguerrera replied:
Resolving since out of scope. It might make sense to move this discussion onto #95

rwoll commented on BuildAndTest.cmd:
```diff
@@ -4,10 +4,6 @@ setlocal
-set RESULT=%ERRORLEVEL%
-if %RESULT% neq 0 goto :exit
-
 call powershell -ExecutionPolicy RemoteSigned -File "%~dp0scripts\BuildAndTest.ps1" %*
```
Removal of this was intentional, correct? Now we only build in CI/CD or if someone explicitly runs BuildAndTestRust.cmd?

I'm good with that since it's no longer a dep of the C# project.

nguerrera replied:
Yes, exactly right.

rwoll commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+
+/// <summary>
+/// A pattern that <see cref="HighPerformanceScanner"/> can scan for. These are
+/// typically the precomputed set derived from the patterns that implement <see
```
```suggestion
/// typically use the precomputed set derived from the patterns that implement <see
```

nit…I think i've fixed it? #Resolved

nguerrera replied:
I've changed it. That edit doesn't reflect what I wanted to say.

rwoll commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// location. This can allow regexes to be shared that would differ only by
+    /// signature.
+    /// </remarks>
+    public Regex Regex { get; }
```
Given the long comment, consider renaming this to `PreSignatureScopedRegex` or something more specific than `Regex`.

I think that more clearly communicates its usage in contexts where the docstring is not immediately presented to you.

rwoll commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+        char s2 = signature[2];
+        char s3 = signature.Length < 4 ? '\0' : signature[3];
+
+        if (s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F)
```
nit: can `Char.IsAscii` be used instead? Or are there perf costs?

The ArgumentException is self-documenting what this `> 0x7F` is doing, so it's already readable w.o the `Char.isAscii` FWIW. #Closed

nguerrera replied:
Yes, but it's .NET 6.0+ only. 😔

michaelcfanning commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
 
+# UNRELEASED
+- DEP: `System.Buffers` dependency has been removed. The package no longer has any dependencies.
+- BRK: .NET 6.0 and 7.0 are no longer supported as they have reached end-of-life. Use a [supported version version of .NET](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core): .NET 8.0 or greater. 
```
>.NET 8.0 or greater [](http://example.com/codeflow?start=197&length=19)

This guidance won't age, i.e., at some point .NET 8.0 itself won't be recommended. So I suggest removing this and just leaving the stable link.

If it was important to leave this marker, I might add something like `as of 2/20/2025` or something like that (and the freshness of the information would be inferable from the release date of the package). But honestly, just zap it I think.

rwoll commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/Azure32ByteIdentifiableKey.cs:
```diff
@@ -16,5 +14,13 @@ public override string Pattern
         }
+
+        private protected override IEnumerable<HighPerformancePattern> HighPerformancePatterns => [
+            new(Signatures.First(),
```
Which sig will this resolve to? And/or why is it safe to assume First is always ok here?

nguerrera replied:
I believe there's an invariant that every IdentifiableKey has only 1 signature. RegexNormalizedSignature above also does Signature.First(). I can codify this in the base class.

michaelcfanning replied:
Yes. For this class of rule this data comprises the provider fixed signature, which is consistent across all secret kinds that are emitted. So multiple rules may share a common provider sig, but each identifiable rule will only be associated with a single (provider) signature.

nguerrera replied:
I have replaced all Signatures.First() including existing ones to go through a signle Signatures.Single() with a comment.

michaelcfanning commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
+- DEP: `System.Buffers` dependency has been removed. The package no longer has any dependencies.
+- BRK: .NET 6.0 and 7.0 are no longer supported as they have reached end-of-life. Use a [supported version version of .NET](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core): .NET 8.0 or greater. 
+- BRK:  .NET Framework 4.5.1 through 4.6.0 are no longer supported. Use a version of .NET Framework version that supports [.NET Standard 2.0](https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0): NET 4.6.1 or greater with .NET 4.7.2 or greater strongly recommended. Note that there are no supported versions of Windows that have a version of .NET Framework that woul ...
+- BRK: `IdentifiableScan` no longer supports stream input.
```
>IdentifiableScan [](http://example.com/codeflow?start=8&length=16)

Is there any API surface area change that came with this? I can answer my own question in a moment, of course, when I look at that file. :)

michaelcfanning replied:
OK, I answered it. :) So things like `IdentifiableScan(byte[], long) have gone bye-bye. It's a good practice to exhaustively describe all the public API we remove explicitly in our release notes.

As `BRK` notes

michaelcfanning commented on docs/ReleaseHistory.md:
```diff
@@ -11,6 +11,16 @@
+- BRK: `IdentifiableScan` no longer supports stream input.
+- BUG: `SecretMasker` and `IdentifiableScan` considered non-alphanumeric delimiter preceding secret to be part of the match for `SEC101/200.CommonAnnotatedSecurityKey`.
+- BUG  `SEC101/061.LooseOAuth2BearerToken` had incorrect signatures, causing no matches to be found unless the input happened to also contain `sig=` or `ret=`
+- PRF: `IdentifiableScan` did not use high-performance scanning techniques for `SEC101/178.AzureIotHubIdentifiableKey` and `SEC101/200.CommonAnnotatedSecurityKey`. This was not intended, but a bug triggered fall-back to slower scanning due to incorrect signatures being used to determine which patterns could be matched with high performance.
```
>fall-back [](http://example.com/codeflow?start=207&length=9)

not a hyphenated term, just use `fallback

michaelcfanning replied:
>to determine which patterns could be matched with high performance [](http://example.com/codeflow?start=275&length=66)

All this can be deleted, it follows naturally from the preceding description, IMO.

`to determine which patterns could be matched with high performance

rwoll commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_156_AzureClientAppIdentifiableCredentials.cs:
```diff
@@ -31,5 +31,18 @@ public override IEnumerable<string> GenerateTruePositiveExamples()
+
+        IEnumerable<HighPerformancePattern> IHighPerformanceScannableKey.HighPerformancePatterns => [
+            new("7Q~",
+                MakeHighPerformancePattern(Pattern.Replace("31,34", "31").Replace("(7|8)", "7"), "7Q~"),
```
TBH, I find this really hard to read with the `.Replace`, and do worry it'll break if someone changes `Pattern` above without realizing it's being used like this below.

I'm not sure what the answer is here, but food for thought.

Do the resolved high perf patterns get emitted to output? Is so, I think that assuages my conerns as a diff would surface that way.

nguerrera replied:
Yes, they get emitted. Look at HighPerformanceScanner.Generated.cs.

I went on a lot of tangents before landing on this approach. More on this to come on another thread.

nguerrera replied:
OK, the "other thread" was the overall comment that doesn't support threading, so I'll continue here.

> At a higher level, though: now each detection has two sets of regexes/patterns: the legacy ones and the high performance ones. Can we avoid this duplication and — when there's high perf versions — exclusively author those and then populate the metadata for the "legacy" with those resolved?

Yes, ideally, there would be an inversion where you built up the general-purpose regex from the simpler scoped one rather than vice versa. I spent a long time trying to do it that way and kept bumping into distracting issues. I would like to do some refactoring of this hierarchy before I attempt that again. This change was large enough that trying to fix that in the same change would have been a bad idea.

Currently, there is an uncomfortable mix of setters and virtual getters. When you move things around, you get into annoying ordering issues. IMHO, neither of these are optimal for what is effectively a bunch of constant data and the mix of them is even worse. There are also efficiency issues here. We have, for example `Pattern => $"..."`. This formats and allocates a new string everytime you do `GetOrCreateRegex(foo.Pattern)`. (And GetOrCreateRegex allocates some more, but that's separate.) This shows up as significant in perf traces. When you try to fix this, you get burned by Pattern depending on other such properties. 

Consider that before this, our dependency on rust was the same amount of duplication. Rust has manually hand-written routines that are equivalent to the "high performance" regexes. The addition of `(9|D|H)` in the rust file in this change shows how they can drift. Given that, IMHO, this is strictly a step forward.

rwoll replied:
ACK. Well-considered and def agree we can address as a follow up during incremental refactors.

nguerrera replied:
#145 for refactoring I'd like to see first before inserting myself into the middle of these property chains.

nguerrera replied:
I will file another issue dependent on that to track the follow-up

nguerrera replied:
#151 

I also removed some DRY for this special case of two signatures in one rule and I think it should be easier to read in the meantime. I've also made it clearer that this code is only used at compile-time, and shaved some bytes of the release build in the process.

michaelcfanning commented on docs/ReleaseHistory_Rust.md:
```diff
@@ -13,10 +13,14 @@ This document contains release notes pertaining to the Rust crates.
 
 # 1.5.3 - 07/26/2024
-- DEP: `System.Text.Json` updated to `v8.0.4` to resolve Depandabot alert.
+- DEP: `System.Text.Json` updated to `v8.0.4` to resolve Dependabot alert.
```
>Dependabot [](http://example.com/codeflow?start=57&length=10)

I kinda like the project name `Depandabot`

michaelcfanning commented on GeneratedRegexPatterns/HighConfidenceSecurityModels.json:
```diff
@@ -361,7 +361,7 @@
   },
   {
-    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?P<secret>secret_scanning_ab85fc6f8d7638cf1c11da812da308d43_[0-9A-Za-z]{5})([^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]|$)",
+    "Pattern": "(^|[^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890])(?P<refine>secret_scanning_ab85fc6f8d7638cf1c11da812da308d43_[0-9A-Za-z]{5})([^abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890]|$)",
```
>refine [](http://example.com/codeflow?start=89&length=6)

there is probably an invariant test somewhere to author to avoid this kind of error moving forward, not a blocking comment. #Resolved

nguerrera replied:
I think that would be cool. I also think this is a good opportunity to more formally discuss a larger testing issue that keeps coming up. I filed #147.

rwoll reviewed (commented):
Finished with my first pass. Overall looking good. I need to go back and pore over some of the core low-level bits again for correctness and my deeper understanding.

At a higher level, though: now each detection has two sets of regexes/patterns: the legacy ones and the high performance ones. Can we avoid this duplication and — when there's high perf versions — exclusively author those and then populate the metadata for the "legacy" with those resolved?

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
I don't like our header formatting. I'd rather is was more columnized. If we can land on something, it should be easy to update a template somewhere and autofix this non-blocking note.

// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT
// license. See LICENSE file in the project root for full license information.
 #WontFix

nguerrera replied:
This is unrelated to this change and follows not only our existing standard, but the existing standard in countless Microsoft repos.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+using System.Buffers;
+using StringInput = System.ReadOnlySpan<char>;
+#else
+using StringInput = string;
```
>StringInput [](http://example.com/codeflow?start=6&length=11)

just asking the question, if we took a conditional dependency on System.Buffers for the netstandard build, would our ability to consume ReadOnlySpan<char> provide any worthwhile perf gains? #Resolved

nguerrera replied:
There's no netstandard API to match a regex on a span so I ended up having to do it with indices, so no.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+internal readonly record struct HighPerformanceDetection
+{
+    /// <summary>
+    /// The signature that identifies the pattern detected.
```
>detected [](http://example.com/codeflow?start=50&length=8)

Strictly speaking the signature is not an identifier for the pattern, these signatures in fact may be shared between patterns.

The signature is fixed data that is guaranteed to be present in any finding and which can therefore be used to perform high-performance look-ups before taking on more expensive regex searching operations.

I think it may be useful to encode some information on useful constraints/limits on signatures. e.g., what minimal length is required to realize meaningful gains (there isn't a precise answer here but empirically in other contexts at least four characters seems required for gains, though 3 chars can be useful if the data pattern is obscure). we typically do not extend signatures beyond 6 chars or so in other contexts.

another thing is that it's helpful if signatures can be shared across multiple patterns wherever possible, so `ABCDEF`, and `ABCDEG` may be useful to collapse into their shared 5-letter component.

I can appreciate this note might not be entirely relevant to this literal expression of signature in the high performance detection record struct. Grab what you think is best and maybe we push more exhaustive stuff elsewhere as appropriate.

nguerrera replied:
I'm going to leave this alone. In the contxt of the high-performance detection, this comment is accurate. More information about signature generally could go elsewhere, but I'll take that as out-of-scope.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    public int Start { get; }
+
+    /// <summary>
+    /// The length of the  in the input that matched in the
```
>The length of the  in the input that matched in the [](http://example.com/codeflow?start=8&length=51)

>The length of the  in the input that matched in the [](http://example.com/codeflow?start=8&length=51)

this sentence needs attention. :)

`The length of the match that was detected in the input.`

or just 
`The length of the detected pattern match.` #Closed

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+{
+    /// <summary>
+    /// The signature that identifies this pattern. This is searched for as fast
+    /// as possible in the input before applying the regex.
```
>regex [](http://example.com/codeflow?start=53&length=5)

This is subtle, but I'd make clear that a search for these must be satisfied, otherwise the regex is not applied.

    /// The signature that identifies this pattern. This data must be found in the input
    /// using a fast, initial search operation; otherwise the regex will not be applied.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// <summary>
+    /// The fixed number of characters that appear before the signature in a match.
+    /// </summary>
+    public int PreSignatureLength { get; }
```
>PreSignatureLength [](http://example.com/codeflow?start=15&length=18)

I can't believe I'm talking about this but strictly speaking `Pre` isn't a word. So this casing is strange from Krzysztof Cwalina's view.

Howzabout `SignaturePrefixLength`

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// <summary>
+    /// The minimum number of characters that can appear in a match.
+    /// </summary>
+    public int MinLength { get; }
```
>MinLength [](http://example.com/codeflow?start=15&length=9)

I'd use `MatchLength` as a term here, minor but helpful clarification.

`MinimumMatchLength` or `Min` pick your poison. Same note for max prop.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    public int MaxLength { get; }
+
+    /// <summary>
+    /// The regular expression that matches the pattern near a signature.
```
>near a signature [](http://example.com/codeflow?start=56&length=16)

>near a signature [](http://example.com/codeflow?start=56&length=16)

Hm, strictly speaking our current rule set for the most precise patterns incorporate a signature. But in some circumstances (like the `sig=` case), the signature is elided at match refinement. So ok.

The pedant in me says use this:
`matches the pattern near or which incorporates a signature`
 #Resolved

nguerrera replied:
That feels awkard to me. I don't think this distinction matters given here all the copious explanation of how this works everywhere near here.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// The whole expression must correspond to the secret. "refine" capture
+    /// group is not used.
+    ///
+    /// The regex msut also be anchored to the beginning of the input.
```
>msut [](http://example.com/codeflow?start=18&length=4)

>msut [](http://example.com/codeflow?start=18&length=4)

`must` #Closed

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// can be applied to find all matches in an arbitrary input. This regex is
+    /// not run on the entire input but only on a substring from <see
+    /// cref="PreSignatureLength" /> before a signature to <see cref="MaxLength"
+    /// /> from there."
```
end of this sentence could be clearer. #Closed

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+
+    /// <summary>
+    /// A rendering of the signature for fast, allocation-free lookup. It is 4
+    /// ascii characters packed into a 32-bit integer.
```
>ascii [](http://example.com/codeflow?start=8&length=5)

>ascii [](http://example.com/codeflow?start=8&length=5)

NIT: use `ASCII` casing #Closed

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+
+    /// <summary>
+    /// A rendering of the signature for fast, allocation-free lookup. It is 4
+    /// ascii characters packed into a 32-bit integer.
```
>packed into a 32-bit integer [](http://example.com/codeflow?start=25&length=28)

do we need a note on endianness here?

nguerrera replied:
I will note the byte order, but also note that it is an implementation detail.

michaelcfanning replied:
ok, so since this is computed on init only, with no other extraction/computation, you are saying it is not important to document the approach, yes?

nguerrera replied:
Yes, exactly.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched befor the regex is applied.
```
>befor [](http://example.com/codeflow?start=48&length=5)

`before`

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
+    /// Framework, even for the unoptimized 'SecretMasker' using only the
+    /// hihghly-identifiable patterns, but this merits more scrutiny and
```
>hihghly [](http://example.com/codeflow?start=8&length=7)

>hihghly [](http://example.com/codeflow?start=8&length=7)

`highly`. #Resolved

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+    ///
+    /// Even more critically, the regex engine abstraction does not currently
+    /// provide a mechanism to search within a substring without allocation.
+    /// there is no incentive to add it, which would require work in the RE2
```
>there [](http://example.com/codeflow?start=8&length=5)

s/be cased `There` NIT

michaelcfanning commented:
>             int iterations = 1000;

for god's sake, drop this value to 1.

---
Refers to: src/Tests.Microsoft.Security.Utilities.Core/IdentifiableScanTests.cs:50 in 7cbf96a. [](commit_id = 7cbf96a3641ca8e5208c94ff2119b7b8ce288bd8, deletion_comment = False)

michaelcfanning commented on src/Tests.Microsoft.Security.Utilities.Core/IdentifiableScanTests.cs:
```diff
@@ -66,10 +65,16 @@ public void IdentifiableScan_IdentifiableKeys()
-                                                                                     signature,
-                                                                                     identifiable.EncodeForUrl);
+                            // Special case: Azure Search identifiable keys must be base62 encoded.
+                            bool isBase62 = signature == IdentifiableMetadata.AzureSearchSignature;
```
>AzureSearchSignature [](http://example.com/codeflow?start=78&length=20)

there are other weirdos out there.

azure functions supports a forward slash but no `+` sign. the reason is that, strictly speaking, you don't need to encode a forward slash in a URL query string (where azure function keys are transported) while the plus sign *must* be encoded. 

if you are *not* seeing breaks from this test, running 1000 times, that may indicate that our test generation for azure functions is producing a too-permissive alphabet in the synthetic data.

nguerrera replied:
I am not seeing it fail, and from inspection, I can see why. The function pattern accepts any base64url:

https://github.com/microsoft/security-utilities/blob/4de8259b8901c5e652a937ba9622173440cce8a8/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_158_AzureFunctionIdentifiableKey.cs#L30

Rust is even more permissive, it accepts the union of base64 and base64url for all HISv1 patterns, which is how azure search was working until now when we stopped using rust.

https://github.com/microsoft/security-utilities/blob/29626a28388b862ba051f7efc320002195b9b730/src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs#L40-L45

I am now actually wondering if it's necessary for the patterns to be as restrictive as the restrictions imposed at generation time. If somebody has an HISv1 key with everything else matching including checksum, but using the wrong alphabet, it wouldn't be much of a false positive to flag it, right? Where the heck did they get that from? The only way that I can think of is if someone removed a spin loop from generation without notifying us.

michaelcfanning replied:
ok, so I was correct about the original azure function implementation (which converted the `+` to `a`) but when they updated their format, they simply converted to URL safe. [https://github.com/Azure/azure-functions-host/blob/5a873034be67cb2300f008bb39ef75fbc9e8b63e/src/WebJobs.Script.WebHost/Security/SecretGenerator.cs#L46](https://github.com/Azure/azure-functions-host/blob/5a873034be67cb2300f008bb39ef75fbc9e8b63e/src/WebJobs.Script.WebHost/Security/SecretGenerator.cs#L46)

nguerrera replied:
OK, that explains that, but what do you think about the slight difference that remains between rust and C# here?

Should we (A) change the rust to be as precise as C#, (B) change C# to be as loose as rust, or (C) leave it alone. I will file a follow-up issue on this.

nguerrera replied:
#159

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/IdentifiableKey.cs:
```diff
@@ -25,6 +25,9 @@ public IdentifiableKey()
         public abstract IEnumerable<ulong> ChecksumSeeds { get; }
 
+        IEnumerable<HighPerformancePattern> IHighPerformanceScannableKey.HighPerformancePatterns => HighPerformancePatterns;
+        private protected abstract IEnumerable<HighPerformancePattern> HighPerformancePatterns { get; }
```
>private protected [](http://example.com/codeflow?start=8&length=17)

what the heck? FamAndAssembly now supported??

nguerrera replied:
Yes.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.Generated.cs:
```diff
@@ -0,0 +1,78 @@
+    /*lang=regex*/ private const string RawRegex0 = """^[a-zA-Z0-9]{52}.{4}9(?:9|D|H)[a-zA-Z0-9][A-L][a-zA-Z0-9]{16}[A-Za-z][a-zA-Z0-9]{7}(?:[a-zA-Z0-9]{2}==)?""";
+    /*lang=regex*/ private const string RawRegex1 = """^[~.a-zA-Z0-9_\-]{3}.{3}[~.a-zA-Z0-9_\-]{31}""";
+    /*lang=regex*/ private const string RawRegex2 = """^[~.a-zA-Z0-9_\-]{3}.{3}[~.a-zA-Z0-9_\-]{34}""";
+    /*lang=regex*/ private const string RawRegex3 = """^[a-zA-Z0-9_\-]{44}.{4}[a-zA-Z0-9_\-]{5}[AQgw]==""";
```
>"""^[ [](http://example.com/codeflow?start=52&length=5)

i love you raw string literals, don't ever change

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.Generated.tt:
```diff
@@ -0,0 +1,67 @@
+
+<#= GenerateCode() #>
+<#+
+private string GenerateCode()
```
>GenerateCode [](http://example.com/codeflow?start=15&length=12)

if we're headed this direction we could unify the generated JSON patterns as well. this would eliminate the post-processing step we do today (as a discrete rules export command).

nguerrera replied:
I really wanted to use Roslyn generator for this, but got very disappointed that I can't chain regex constant generation to regex code generation. Maybe it's possible to wrap the roslyn generator and use a different attribute or something, though, I will file an issue on it.

I'm not happy with T4 here, but I would also have been mad at myself if I made the problem of rust getting out of sync with C# on these patterns even worse with a third copy.

nguerrera replied:
#152

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -2,141 +2,64 @@
             foreach (string signature in pattern.Signatures)
             {
-                if (!HighPerformanceEnabledSignatures.Contains(signature))
+                var highPerformancePattern = CompiledHighPerformancePattern.ForSignature(signature);
```
>CompiledHighPerformancePattern [](http://example.com/codeflow?start=45&length=30)

i'm surprised `var` is permitted here as the expressed type of the local isn't apparent from the remaining code pattern.

nguerrera replied:
It's because I deleted stylecop which broke on new patterns. I'll put the enforcement back with dotnet format. I'll make a different PR for that. See also other comment on var, which was meant to be my last, but then I replied on this related detail.

nguerrera replied:
Actually maybe not why because I see some other pre-exiting vars that don't have obvious type. But regardless, once I match cask formatting check here, this will be enforced.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -146,268 +69,68 @@ private void PopulateBackupMasker(bool generateCorrelatingIds)
+    public IEnumerable<Detection> DetectSecrets(string input)
     {
-        get
+        foreach (var hpDetection in highPerformanceSecretScanner.Scan(input))
```
>hpDetection [](http://example.com/codeflow?start=21&length=11)

not a fan of this var use here, had to dig more to know the type.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -146,268 +69,68 @@ private void PopulateBackupMasker(bool generateCorrelatingIds)
         {
-            if (this.scan != IntPtr.Zero)
+            string found = input.Substring(hpDetection.Start, hpDetection.Length);
+            foreach (var pattern in this.signatureToPatternsMap[hpDetection.Signature])
```
>var [](http://example.com/codeflow?start=21&length=3)

i'll stop commenting on this but personally i only prefer `var` when the type is clear, because of an explicit cast or an explicit reference via a type in a `new` construction.

nguerrera replied:
I actually prefer to `var` all the things. Do it for a while and it's really hard to go back. A lot of modern languages have pervasive type inference all over the place and it's glorious and liberating in my experience. Really hard to come back to type names all over the place after my a hiatus from .NET. 

One thing I never understood is why people are fine not seeing the type of expressions that aren't assigned to anything like `Foo(Bar())` is fine, `var bar = Bar(); Foo();` is terrible. I don't get it, but this is the last I'm going to speak of this, lol. I will comply.

This one is trivially configurable in dotnet format so that makes me very happy. I can use `var` to my heart's content and then let it fix it for me. I actually made sure cask was configured like that and I'll do that here too.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,505 @@
+/// <summary>
+/// A detection from <see cref="HighPerformanceScanner"/>.
+/// </summary>
+internal readonly record struct HighPerformanceDetection
```
>HighPerformanceDetection [](http://example.com/codeflow?start=32&length=24)

>HighPerformanceDetection [](http://example.com/codeflow?start=32&length=24)

NIT: what's your care factor around the `single type, single file` standard? i usually conform to this. #Resolved

nguerrera replied:
I don't like the standard, but my care factor is not high enough to push back.

nguerrera replied:
Again, my time away from .NET influenced me here.

nguerrera replied:
Done. Thank you, refactoring -> move to file. I will do it pre-checkin next time. :)

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -146,268 +69,68 @@ private void PopulateBackupMasker(bool generateCorrelatingIds)
+            foreach (var pattern in this.signatureToPatternsMap[hpDetection.Signature])
             {
-                return identifiable_scan_match_count(this.scan);
+                Detection detection = ProcessDetection(hpDetection, found, pattern);
```
>ProcessDetection [](http://example.com/codeflow?start=38&length=16)

>ProcessDetection [](http://example.com/codeflow?start=38&length=16)

So this name is super generic. You didn't create this problem. Maybe we could put some thought into it. `FinalizeDetection` might be better, this code calls into a refinement phase that itself might return null (in which case the detection is dropped entirely) #Closed

nguerrera replied:
Yes, I wasn't sure what to name this. I like `FinalizeDetection` better so I'll go with it. If you think of something even better, please feel free to change it.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/RegexPattern.cs:
```diff
@@ -358,4 +358,50 @@ internal static string NormalizeGroupsPattern(string pattern)
+            pattern = pattern.Substring(refineStart.Length, pattern.Length - refineStart.Length - 1);
+        }
+
+        pattern = pattern.Replace(WellKnownRegexPatterns.Uppercase, "A-Z");
```
>WellKnownRegexPatterns.Uppercase [](http://example.com/codeflow?start=34&length=32)

we should push this improvement everywhere, as you filed an issue against.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/IHighPerformanceScannableKey.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        public string Signature { get; }
+        public string Regex { get; set; }
+        public int PreSignatureLength { get; }
```
>PreSignatureLength [](http://example.com/codeflow?start=19&length=18)

>PreSignatureLength [](http://example.com/codeflow?start=19&length=18)

hoping for a rename here as per other note. we don't need to block on it. #Closed

michaelcfanning reviewed (dismissed):
Marking this as approved, after adding many NITS, some rename suggestions. I think that some of these changes aren't a great fit for our current design but resolving this is more contigent on changing existing design changes than changing what you've done to bolt this on. So, imo, we should get this in as soon as we can, start knocking off some of the separate improvements you cited, then take another look at a more holistic approach to the new scanning approach.

rwoll commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/IdentifiableKey.cs:
```diff
@@ -5,26 +5,33 @@
+        // Identifiable key patterns must have exactly one signature.
+        internal string Signature => Signatures!.Single();
+
+        public string RegexNormalizedSignature => Signature.Replace("+", "\\+");
```
1. Consider using net451-compatible [`Regex.Escape(…)`](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.escape?view=net-9.0). Took me a few reads to understand the intent of the `.Replace` in the proposed code…wich leads me to my next suggestion…
2. `Normalized` is kinda ambiguous. Consider renaming it to `RegexEscapedSignature` or `RegexSafeSignature`.

Once I see it used in context, all the names work, but looking at it outside that context it's not clear. If this was intental, I'd care less but since it's a `public` worth a consideration.

I could go either way on all fronts, so consider the entirety of this comment a nit :)

rwoll replied:
And another food for thought: why do we escape it here? It may make more sense to escape it just before slotting it into the regex.

nguerrera replied:
Fully agree, but note that this is all pre-existing.

The diff is rendering sub-optimally. This is how I would have wanted it to show up:
```diff
+   // Identifiable key patterns must have exactly one signature.
+   internal string Signature => Signatures!.Single();

< public string RegexNormalizedSignature => Signatures!.First().Replace("+", "\\+");
> public string RegexNormalizedSignature => Signature.Replace("+", "\\+");
```
Basically I just added `Signature` with a comment and replaced all the Signatures.First() with Signature. I felt this address the previous PR note. 

I will use Regex.Escape, but I'm not going to rename existing public things in this already massive PR.

nguerrera committed:
Implement high-performance identifiable key scanning in managed code

This change removes the interop with the rust API from `IdentifiableScan` and implements similar high-performance scanning techniques in managed code. This reduces the size of the NuGet package tremendously, from ~7 MB to ~200 KB, and adds support for high-performance scanning on non x86/x64 CPUs and non-Windows operating systems.

Performance is generally better than before for managed callers of `IdentifiableScan` with significant gains in most cases. See benchmarks below.

Code generation is used to avoid having to keep the high-performance scanning code in sync when patterns are changed. See also extensive code comments on the approach.

For now, the streaming capability of `IdentifiableScan` has been removed, but this can be re-added as the managed algorithm is actually streaming friendly by only ever regex-matching against short sub-ranges of input. This was held back for lack of time and discovery that our only partner that calls this API does not use the streaming.

Finally, the `StyleCop` analyzer is removed because it broke on newer C# constructs. This will be replaced with a `dotnet format` check like the one in `microsoft/cask` in a follow-up change.

nguerrera commented:
@michaelcfanning @rwoll Please take another look. I have:

- addressed all comments
- fixed up issues with benchmark
- put benchmark results and more details in description
- added some tests

I have changed enough here that a full review pass over the entire change is warranted.

nguerrera committed:
Merge branch 'main' into managed-identifiable-scan

nguerrera commented:
I will need to adjust the PR validation settings on GitHub to expect the debug and release .NET runs. I will wait until this is approved and do it right before merging. Otherwise, I will break other PRs.

rwoll commented on GeneratedRegexPatterns/PreciselyClassifiedSecurityKeys.json:
```diff
@@ -4,7 +4,7 @@
     "Id": "SEC101/200",
     "Name": "CommonAnnotatedSecurityKey",
     "Signatures": [
-      "JQQJ9"
```
Prior to merging, please double check these JSON files are the most up-to-date generated versions. We have a logged issue to automate this check, but for now, this comment serves as the reminder to run the command.

nguerrera replied:
Double checked.

nguerrera replied:
Even better, approve #162 to check.

rwoll commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,274 @@
+    /// <summary>
+    /// Gets the pre-computed pattern for the given signature or null if there is none.
+    /// </summary>
+    public static CompiledHighPerformancePattern ForSignature(string signature)
```
```suggestion
    public static CompiledHighPerformancePattern? ForSignature(string signature)
```

rwoll commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,274 @@
+    /// iteration. If more complex regexes are added, the generated code should
+    /// be re-examined.
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched before the regex is applied.
+    ///
```
Is there an explicit test covering this `SingleLine` behavior?

nguerrera replied:
Hmm, there is not and it's hard to test without a very silly test. This is purely an optimization. The patterns have no `.` wildcards other than in the signature but the regex is only run if the signature was found. So I can't check if the `.` matches `\n`. 

I also can't test with different regex either like the other unit tests because that wouldn't get code gen'ed to construct the Regex with these options.

nguerrera replied:
I thought of a test that's not too silly. Working on it.

nguerrera replied:
...And it caught a different bug, we're not actually sharing the regex instances on .NET Framework :) Fixing. Thank you for pressing me on this.

nguerrera replied:
Fixed.

rwoll commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,274 @@
+
+        foreach (IHighPerformanceScannableKey generalPattern in WellKnownRegexPatterns.HighConfidenceMicrosoftSecurityModels.OfType<IHighPerformanceScannableKey>())
+        {
+            foreach (HighPerformancePattern pattern in generalPattern.HighPerformancePatterns)
```
This'll be deterministic ordering, correct?

nguerrera replied:
Correct.

rwoll commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,274 @@
+
+        foreach (KeyValuePair<string, int> regex in regexes.OrderBy(p => p.Value))
+        {
+            sb.AppendLine($""""    /*lang=regex*/ private const string RawRegex{regex.Value} = """{regex.Key}""";"""");
```
TIL about the editor lang hints. Neat!

rwoll commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,230 @@
+/// </summary>
+internal sealed class HighPerformanceScanner
+{
+#if NET9_0_OR_GREATER
+    private SearchValues<string> _signatures;
+#elif NET8_0_OR_GREATER
+    private SearchValues<char> _signatureStarts;
+#else
+    private char[] _signatureStarts;
+#endif
```
yay for supporting all. the. things! #Resolved

rwoll commented on src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/Azure64ByteIdentifiableKey.cs:
```diff
@@ -18,5 +17,15 @@ public override string Pattern
+        private protected override IEnumerable<HighPerformancePattern> HighPerformancePatterns => [
+            new(Signature,
+                MakeHighPerformancePattern(Pattern, Signature),
+                signaturePrefixLength: 76,
+                minMatchLength: 88,
+                maxMatchLength: 88)
```
I trust all these numbers are correct based on the tests passing. :)

rwoll reviewed (dismissed):
Just some nits. Nice. work. this. is. epic!

Your PR description will all the benchmarks also rocks! ⚡️

nguerrera committed:
PR feedback: nullable

nguerrera committed:
PR feedback: Test and fix regex sharing

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+    /// not run on the entire input but only on a substring determined by <see
+    /// cref="SignaturePrefixLength"/> and <see cref="MaxMatchLength"/>.
+    ///
+    /// Example: Let S be a signature character, X be a character--xpected to
```
>xpected [](http://example.com/codeflow?start=67&length=7)

` expected`? NIT

nguerrera replied:
Sigh, I didn't even finish the example. I'm just going to delete this paragraph for now.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+    /// not run on the entire input but only on a substring determined by <see
+    /// cref="SignaturePrefixLength"/> and <see cref="MaxMatchLength"/>.
+    ///
+    /// Example: Let S be a signature character, X be a character--xpected to
```
>xpected [](http://example.com/codeflow?start=67&length=7)

vertical whitespace here to remove

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+    ///
+    /// Also note that the high-performance scanner does not allow substitution
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
```
>Profiling showed no performance gain from RE2 [](http://example.com/codeflow?start=14&length=45)

we know this statement is suspect due to bugs in our impl but will revisit later

nguerrera replied:
Yes, more details in #158. Since this comment also says "this merits more scrutiny" and we are doing that as we speak, I don't think we need to change the comment.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+
+
+    public CompiledHighPerformancePattern(string signature,
+                                           int signaturePrefixLength,
```
>int [](http://example.com/codeflow?start=43&length=3)

NIT parameters not in alignment

nguerrera replied:
Because of refactor -> rename, no doubt. 😅

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+                                           Regex scopedRegex)
+    {
+#if HIGH_PERFORMANCE_CODEGEN // This validation happens at compile time and needn't be repeated at runtime.
+        if (!scopedRegex.ToString().StartsWith("^"))
```
>scopedRegex [](http://example.com/codeflow?start=13&length=11)

why is this important? nice to comment on in code later. also, the 3 or 4 char limit, which would 5 chars be unhelpful for example. note that other projects declare lengthier signatures for general IndexOf check, will we need limited fix sig expression to generalize this class for those scenarios?

nguerrera replied:
Both of these can be relaxed, and I'm sure they will be when I look at #156.

I didn't want the cost of doing things that rust identifiable_scan does not do to show up in the benchmark here.

nguerrera replied:
More details.

If the regexes are not `^` anchored then they cannot have a fixed signature-prefix-length. I suspect I will change this to *max*-signature-prefix-length and only the rules that actually have fixed length will use `^`.

If rules have variable pre-signature length, then we also can't play the trick of using the same regex that skips over the signature.

If I try to explain all this in a comment, I will surely make a bunch more typos, so I'll do it in a follow-up.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+        char s3 = signature.Length < 4 ? '\0' : signature[3];
+
+#if HIGH_PERFORMANCE_CODEGEN
+        if (s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F)
```
>s0 > 0x7F || s1 > 0x7F || s2 > 0x7F || s3 > 0x7F [](http://example.com/codeflow?start=12&length=48)

seems like someone somewhere someday might also want this helper, as some inlinable utility

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+            {
+            """);
+
+        // Since patterns can share regexes, we deduplicate them and assign an ID using this counter.
```
>regexes [](http://example.com/codeflow?start=36&length=7)

in practice it's the scoped regexes, yes, that are duplicated? at a high level we expect uniqueness to complete regexes at the rule level.

nguerrera replied:
I think this is actually true even before "scoping", though "scoping" makes it true in more cases.

For example, consider these two:


https://github.com/microsoft/security-utilities/blob/0b88ee4ee6d368f2fd6607122fddb57144c7d1b3/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_178_AzureIotHubIdentifiableKey.cs#L16

https://github.com/microsoft/security-utilities/blob/0b88ee4ee6d368f2fd6607122fddb57144c7d1b3/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_179_AzureIotDeviceProvisioningIdentifiableKey.cs#L16

They have the same signature and hence the same regex. They differ only in how they post-process a regex match using the differing checksum seeds.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs:
```diff
@@ -0,0 +1,285 @@
+                    regexes[pattern.ScopedRegex] = id;
+                }
+
+                // Generate the consturctor call the pattern.
```
>consturctor [](http://example.com/codeflow?start=32&length=11)

`constructor`

nguerrera replied:
```suggestion
                // Generate the call to the pattern constructor.
```

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.Generated.tt:
```diff
@@ -0,0 +1,66 @@
+                new CompilerError(Host.TemplateFile, 1, 1, "", $"Error generating code: {ex}")
+            });
+            var sb = new System.Text.StringBuilder();
+            sb.AppendLine("/* ERROR: ");
```
>ERROR [](http://example.com/codeflow?start=30&length=5)

it was not my experience working in your branch that the code was left in a compilable state, but i can appreciate it might be hard to guarantee this. look at this code, it's possible you also made some improvements here i didn't get benefit of.

nguerrera replied:
Yeah, this is not bullet-proof. There are still cases where you have to revert the code to re-generate:

1. An error that causes T4 to crash before it gets inside our `try`.
2. A bug in the generation routine that emits bad code.

I am really hoping I can figure out #152 and this will be short-lived. That said, this helps a lot and it's not *that* bad as-is.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/HighPerformanceScanner.cs:
```diff
@@ -0,0 +1,232 @@
+    /// <summary>
+    /// Scan the input for all patterns and return the detections found.
+    /// </summary>
+    public List<HighPerformanceDetection> Scan(StringInput input)
```
>Scan [](http://example.com/codeflow?start=42&length=4)

super clean, nice

nguerrera committed:
Update src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

nguerrera committed:
PR feedback

michaelcfanning reviewed (approved):
:shipit:

nguerrera commented:
Waiting for macos jobs to run to merge.
